File: fsharp/poker/Poker.fs
module Poker

type Hand = {
    ranks    : int array // sorted in descending order
    flush    : bool 
    straight : bool
}

type Level =  
    | HighCard      = 1 
    | Pair          = 2     
    | TwoPair       = 3 
    | ThreeOfAKind  = 4 
    | Straight      = 5 
    | Flush         = 6 
    | FullHouse     = 7 
    | FourOfAKind   = 8 
    | StraightFlush = 9

let ACE_TO_5 = [|14; 5; 4; 3; 2|] // special straight

let createHand (deal : string) = 
    let (rks, sts) =
        deal.Split ' '
        |> Array.map (fun c -> ("__234567891JQKA".IndexOf(c[0]), c[String.length(c) - 1]))
        |> Array.sortByDescending fst
        |> Array.unzip 
    { ranks    = rks
      flush    = sts |> Array.distinct |> Array.length = 1
      straight = (rks = ACE_TO_5) || (rks |> Array.distinct |> Array.length = 5 && rks[0] - rks[4] = 4) }

let getLevelAndCrucialRanks hand = 
    let (ranksByFreqsDesc, freqs) = 
        hand.ranks
        |> Array.countBy id 
        |> Array.sortByDescending (fun (rk, freq) -> (freq, rk))
        |> Array.unzip
    let lvl = 
        match (Array.length freqs, hand.straight, hand.flush) with 
        | (4, _, _)                   -> Level.Pair
        | (3, _, _) when freqs[1] = 1 -> Level.ThreeOfAKind
        | (3, _, _)                   -> Level.TwoPair
        | (2, _, _) when freqs[0] = 4 -> Level.FourOfAKind
        | (2, _, _)                   -> Level.FullHouse 
        | (5, true , false)           -> Level.Straight
        | (5, false, true )           -> Level.Flush
        | (5, true , true )           -> Level.StraightFlush
        | _                           -> Level.HighCard
    (lvl, ranksByFreqsDesc)

let rec getScore ranksByFreqsDesc =
    if ranksByFreqsDesc = ACE_TO_5 then getScore [|5; 4; 3; 2; 1|] // 5 is highest card
    else ranksByFreqsDesc |> Array.reduce (fun acc el -> acc * 14 + el)

let topGroupBy f =
    List.groupBy f >> List.sortByDescending fst >> List.head >> snd

let bestHands (deals : string list) = 
    deals
    |> List.map (fun deal -> (deal, deal |> createHand |> getLevelAndCrucialRanks))
    |> topGroupBy (snd >> fst) // extract the hands with best level
    |> List.map (fun (deal, (_, rks)) -> (deal, getScore rks)) // throw away level and score the crucial ranks
    |> topGroupBy snd // extract the hands with best score
    |> List.unzip |> fst // extract the strings of these deals

File: csharp/poker/Poker.cs
using System;
using System.Linq;
using System.Collections.Generic;

public static class Poker
{
    // extension method to unzip IEnumerables of tuples (T1, T2) again (here already converted to arrays of the kinds T1 and T2)
    // call it with "var (fst, snd) = EnumerableToUnzip.Unzip(); " 
    private static (T1[], T2[]) Unzip<T1, T2>(this IEnumerable<(T1, T2)> src) => (src.Select(t => t.Item1).ToArray(), src.Select(t => t.Item2).ToArray());
    
    // only have to know the ranks and if it is a flush or a straight, so precalculate these when parsing the deal
    private struct hand { 
        internal int[] ranks; 
        internal bool flush;
        internal bool straight;
        internal hand(string deal)
        {
            // calling the unzip method to get ranks and suits separately (can leave the suits as characters, only interesting for flush)
            var (rks, sts) = deal
                .Split(' ')
                .Select(c => ("__234567891JQKA".IndexOf(c.First()), c.Last()))
                .OrderByDescending(t => t.Item1)
                .Unzip();
            ranks = rks;
            flush = sts.Distinct().Count() == 1;
            straight = ranks.Distinct().Count() == 5 && ranks[0] - ranks[4] == 4;
        }
    }

    public static IEnumerable<string> BestHands(IEnumerable<string> hands)
    {
        var scoredDeals = hands.Select(deal => (GetScore(new hand(deal)), deal));
        var mx = scoredDeals.Unzip().Item1.Max(); // calling Unzip() again
        return scoredDeals.Where(tuple => tuple.Item1 == mx).Select(tuple => tuple.Item2);
    }

    private static long[] FACTORS = new[] { 10_000_000_000, 100_000_000, 1_000_000, 10_000, 100, 1 }; 
    // at most six items (level + possible 5 crucial ranks) are multiplied with factors and summed up to give a comparable score
    private static long GetScore(hand h) =>   
        FACTORS.Zip(GetLevelAndCrucialRanks(h)).Select(pr => pr.First * (long)pr.Second).Sum();
    
    private static int[] ACE_TO_5 = new[] { 14, 5, 4, 3, 2 };
    private static IEnumerable<int> GetLevelAndCrucialRanks(hand h)
    {   
        if (h.ranks.SequenceEqual(ACE_TO_5)) // handle the special straight first
        {
            if (h.flush) return new[] { 8, 5 };
            else return new[] { 4, 5 };
        }
        // again use the Unzip() - method to get two separate arrays   
        var (freqs, ranksByFreqsDesc) = h.ranks
            .GroupBy(r => r)
            .OrderByDescending(g => g.Count())
            .ThenByDescending(g => g.Key)
            .Select(g => (g.Count(), g.Key))
            .Unzip();

        return (h.flush, h.straight, freqs[0]) switch
        {
            (true, true, _)                    => new[] { 8, h.ranks[0] },     // straight flush
            (_   , _   , 4)                    => ranksByFreqsDesc.Prepend(7), // 4 of a kind
            (_   , _   , 3) when freqs[1] == 2 => ranksByFreqsDesc.Prepend(6), // full house
            (true, _   , _)                    => h.ranks.Prepend(5),          // flush
            (_   , true, _)                    => new[] { 4, h.ranks[0] },     // straight
            (_   , _   , 3)                    => ranksByFreqsDesc.Prepend(3), // 3 of a kind
            (_   , _   , 2) when freqs[1] == 2 => ranksByFreqsDesc.Prepend(2), // two pairs
            (_   , _   , 2)                    => ranksByFreqsDesc.Prepend(1), // one pair
            _                                  => h.ranks.Prepend(0)           // high card
        };
    }
}